File: interview/Cracking the Coding Interview/chapter 2/2.4/c++/2.4.cpp
#include <iostream>

using namespace std;

template <typename T>
class Node {
	public:
	Node() { }
	
	Node(T data, Node<T>* next){
		this->data = data;
		this->next = next;
	}
	
	T data;
	Node<T>* next;
};

template <typename T>
class LinkedList {
	public:
	LinkedList () { this->head = 0;}
	void insert(T data){
		Node<T> * n = new Node<T>(data,head);
		head = n;
	}
	
	Node<T>* search(T data) {
		Node<T>* curr = head;
		while (curr && curr->data != data){
			curr = curr->next;
		}
		
		return curr;
	}
	void partition(T x){
		Node<T>* node_x = search(x);
		Node<T>* prev;
		Node<T>* curr = head;
		
		bool past_x = false;
		while (curr){
			if (curr == node_x) {
				past_x = true;
				prev = curr;
				curr = curr->next;
			}
			else {
				if (!past_x && curr->data > node_x->data){
					//move to X
					Node<T>* next = curr->next;
					if (curr == head)
						head = head->next;
					else
						prev->next = next;
					curr->next = node_x->next;
					node_x->next = curr;
					curr = next;
				}
				else if (past_x && curr->data < node_x->data){
					// move to head
					Node<T>* next = curr->next;
					curr->next = head;
					head = curr;
					prev->next = next;
					curr = next;
				}
				else {
					prev = curr;
					curr = curr->next;
				}
			}
				
		}
	}
	
	void print(){
		Node<T>* curr = head;
		while (curr){
			cout << curr->data << " ";
			curr = curr->next;
		}
		cout << endl;
	}
	private:
	Node<T>* head;
};


int main(){
	LinkedList<int> L = LinkedList<int>();
	
	cout << "3 middle of list:\n";
	L.insert(4);
	L.insert(2);
	L.insert(3);
	L.insert(1);
	L.insert(5);
	L.print();
	L.partition(3);
	L.print();
	
	L = LinkedList<int>();
	cout << "3 head of list:\n";
	L.insert(4);
	L.insert(2);
	L.insert(1);
	L.insert(5);
	L.insert(3);
	L.print();
	L.partition(3);
	L.print();
	
	L = LinkedList<int>();
	cout << "3 end of list:\n";
	L.insert(3);
	L.insert(4);
	L.insert(2);
	L.insert(1);
	L.insert(5);
	L.print();
	L.partition(3);
	L.print();
	return 0;
}

File: interview/Cracking the Coding Interview/chapter 2/2.4/python/2.4.py
class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node

    def get_data(self):
        return self.data

    def get_next(self):
        return self.next_node

    def set_next(self, new_next):
        self.next_node = new_next

    def __repr__(self):
        return str(self.data)
    
class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    def insert(self, data):
        new_node = Node(data)
        new_node.set_next(self.head)
        self.head = new_node

    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.get_next()
        return count

    def __len__(self):
        return self.size()

    def search(self, data):
        current = self.head
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                current = current.get_next()
        if current is None:
            raise ValueError("Data not in list")
        return current

    def __contains__(self, data):
        return self.search(data)
        
    def delete(self, data):
        current = self.head
        previous = None
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                previous = current
                current = current.get_next()

        if current is None:
            raise ValueError("Data not in list")
        if previuos is None:
            self.head = current.get_next()
        else:
            previous.set_next(current.get_next())

    def __repr__(self):
        s = ''
        current = self.head
        while current:
            s += str(current.get_data()) + ','
            current = current.get_next()
        return s[:-1]

    def partition(self,x):
        X = self.search(3)
        after_X = False
        prev = None
        i = self.head
        while i != None:
            nxt = i.get_next()
            nxt_prev = i
            if i != X:
                if after_X and i.get_data() < X.get_data():
                    prev.set_next(i.get_next())
                    nxt_prev = i.get_next()
                    i.set_next(self.head)
                    self.head = i
                elif not after_X and i.get_data() > X.get_data():
                    if i == self.head:
                        self.head = nxt
                        nxt_prev = None
                    else:
                        prev.set_next(i.get_next())
                        nxt_prev = i.get_next()
                        
                    i.set_next(X.get_next())
                    X.set_next(i)
                    
                        
            else:
                after_X = True
                
            i = nxt
            prev = nxt_prev

# test when X=3 in middle of list
L = LinkedList(Node(5,Node(1,Node(3,Node(2,Node(4))))))
L.partition(3)
print(L)

# test when X=3 head of list
L = LinkedList(Node(3,Node(2,Node(5,Node(1,Node(4))))))
L.partition(3)
print(L)

# test when X=3 tail of list
L = LinkedList(Node(1,Node(5,Node(2,Node(4,Node(3))))))
L.partition(3)
print(L)
